iXBRLViewerPlugin/iXBRLViewer.py

```diff
@@ -480,32 +490,38 @@ def createViewer(
 
         if packageDownloadURL is not None:
             self.taxonomyData["filingDocuments"] = packageDownloadURL
-        elif len(self.reports) == 1 and os.path.dirname(self.reports[0].modelDocument.filepath).endswith('.zip'):
-            filingDocZipPath = os.path.dirname(self.reports[0].modelDocument.filepath)
-            filingDocZipName = os.path.basename(filingDocZipPath)
-            iv.addFilingDoc(filingDocZipPath)
+        elif self.fromSingleZIP:
+            filingDocZipName = os.path.basename(self.filingDocZipPath)
+            self.iv.addFilingDoc(self.filingDocZipPath)
             self.taxonomyData["filingDocuments"] = filingDocZipName
 
-        if not self.addViewerData(iv.files[0], scriptUrl):
+        if not self.addViewerData(self.iv.files[0], scriptUrl):
             return None
 
-        return iv
+        if len(self.iv.files) == 1:
+            # If there is only a single report, call the output file "xbrlviewer.html"
+            # We should probably preserve the source file extension here.
+            self.iv.files[0].filename = 'xbrlviewer.html'
+
+        return self.iv
 
 
 class iXBRLViewerFile:
 
     def __init__(self, filename, xmlDocument):
         self.filename = filename
-        self.xmlDocument = xmlDocument
+        self.xmlDocument = deepcopy(xmlDocument)
+        # Arelle's custom element class lookup relies on data that is stored on
+        # ModelXbrl, which has been cleaned-up by the time we want to serialize
+        self.xmlDocument.parser.set_element_class_lookup(etree.ElementDefaultClassLookup())
```

Comment:
> Out of curiosity, why is this needed for serialization? I would have guessed that parser.set_element_class_lookup is only used when parsing.

That would have been my expectation too.  It's a call to `getroot()` that triggers it:

```
Exception 'ModelXbrl' object has no attribute 'qnameConcepts'
...
    for child in viewerFile.xmlDocument.getroot():
   File "src/lxml/etree.pyx", line 1219, in lxml.etree._Element.__iter__
   File "src/lxml/etree.pyx", line 2877, in lxml.etree.ElementChildIterator.__cinit__
   File "src/lxml/etree.pyx", line 1636, in lxml.etree._elementFactory
   File "src/lxml/classlookup.pxi", line 403, in lxml.etree._parser_class_lookup
   File "src/lxml/classlookup.pxi", line 456, in lxml.etree._custom_class_lookup
   File "/home/pdw/Arelle/arelle/ModelObjectFactory.py", line 141, in lookup
    return self.modelXbrl.matchSubstitutionGroup(qnameNsLocalName(ns, ln), elementSubstitutionModelClass)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   File "/home/pdw/Arelle/arelle/ModelXbrl.py", line 471, in matchSubstitutionGroup
    elementMdlObj = self.qnameConcepts.get(elementQname)
                    ^^^^^^^^^^^^^^^^^^
```

Looking at the [etree source](https://github.com/lxml/lxml/blob/lxml-5.1/src/lxml/etree.pyx#L2900) I think it's creating Python proxy nodes on-the-fly when needed based on the C nodes (see also [proxy.pxi](https://github.com/lxml/lxml/blob/lxml-5.1/src/lxml/proxy.pxi)), so depending on what's cached, an element class lookup can happen when reading nodes.

> what code is doing the clean-up, and what is it doing? 

I think it's [ModelXbrl.close()](https://github.com/Arelle/Arelle/blob/master/arelle/ModelXbrl.py#L369) and I think it's `self.__dict__.clear()` that kills the `qnameConcepts` dict, which is used when trying to lookup substitution groups when instantiating Python proxies.

> Does "not safe" just mean that the code fails with a NoneError (or whatever)

Yes - the above Python exception.

Having now looked more carefully at the etree/libxml2 source code, I think that `deepcopy` only copies the C nodes and this doesn't include the `_private` value where etree stores its proxy nodes, so doing the deepcopy clears the cache of Python proxy nodes, and once the custom class lookup is removed we're left with a clean lxml.etree tree.

Reply:
Thanks, that helps.  Maybe this comment is clearer?
> deepcopy does not retain the Python proxies, so iterating the node tree must create new ones.  However, the original ModelObjectFactory is still referenced, and that references a potentially closed ModelXbrl, so we must use a different lookup object.  We don't rely on any specific proxy, so just use the default.

That said, if I'm reading the code correctly, it seems like even with this change, we're still using the [stale element factory](https://github.com/lxml/lxml/blob/ad92005a3591b4d0fbe626b28bd68ec692c82351/src/lxml/etree.pyx#L818-L839) for the root element and just getting lucky that the Arelle logic doesn't fail for it?  I guess we could extend the comment to say that.  Alternatively, I guess we could serialize/reparse using etree (unfortunate), or we could avoid deepcopy and manually copy into a custom mini-etree that only supports the features needed by XHTMLSerializer...

Reply:
I don't understand what circumstances the [`if root is not None`](https://github.com/lxml/lxml/blob/ad92005a3591b4d0fbe626b28bd68ec692c82351/src/lxml/etree.pyx#L831) line applies to, but yes, it does seem possible that we're using the Arelle class lookup.

Other options:

1. Have Arelle unset the class lookup on clean up.
2. Detect if the models have been cleaned up, and if so, unset the class lookup before the `deepcopy`
3. Stash and restore the class lookup before/after the `deepcopy`, but this relies on reading the private `_class_lookup` member.

Reply:
> That said, if I'm reading the code correctly, it seems like even with this change, we're still using the [stale element factory](https://github.com/lxml/lxml/blob/ad92005a3591b4d0fbe626b28bd68ec692c82351/src/lxml/etree.pyx#L818-L839) for the root element and just getting lucky that the Arelle logic doesn't fail for it?

Having looked at this again, I think this is actually safe. The element factory isn't stale at the point that we do the copy, because we instantiate the `iXBRLViewerFile` object as part of processing the model.  

The serialization happens later, but by that time we've replaced the class lookup.  It does mean that we potentially serialize an Arelle proxy object for the root, but I think that's safe as the Arelle proxies don't interfere with standard lxml operations.

I've revised the comment in cdd035e

Reply:
That makes sense, thanks.